ProcessLib/SmallDeformationWithLIE/LocalAssembler/SmallDeformationLocalAssemblerMatrixNearFracture-impl.h

```diff
@@ -0,0 +1,228 @@
+                          KelvinVectorDimensions<DisplacementDim>::value);
+
+        _secondary_data.N[ip] = sm.N;
+    }
+}
+
+
+template <typename ShapeFunction, typename IntegrationMethod,
+          int DisplacementDim>
+void SmallDeformationLocalAssemblerMatrixNearFracture<ShapeFunction, IntegrationMethod,
+                                    DisplacementDim>::
+assembleWithJacobian(
+    double const t,
+    Eigen::VectorXd const& local_u,
+    Eigen::VectorXd& local_b,
+    Eigen::MatrixXd& local_J)
+{
+    assert (_element.getDimension() == DisplacementDim);
+
+    auto const n_dof_per_var = ShapeFunction::NPOINTS * DisplacementDim;
+
+    auto localRhs_ru = local_b.segment(0, n_dof_per_var);
+    auto localRhs_du = local_b.segment(n_dof_per_var, n_dof_per_var);
+
+    auto localA_uu = local_J.block(0, 0, n_dof_per_var, n_dof_per_var);
+    auto localA_ug = local_J.block(0, n_dof_per_var, n_dof_per_var, n_dof_per_var);
+    auto localA_gu = local_J.block(n_dof_per_var, 0, n_dof_per_var, n_dof_per_var);
+    auto localA_gg = local_J.block(n_dof_per_var, n_dof_per_var, n_dof_per_var, n_dof_per_var);
+
+    auto const nodal_ru = local_u.segment(0, n_dof_per_var);
+    auto nodal_du = local_u.segment(n_dof_per_var, n_dof_per_var);
+
+    auto const &fracture_props = *_process_data._fracture_property;
+
+    unsigned const n_integration_points =
+        _integration_method.getNumberOfPoints();
+
+    SpatialPosition x_position;
+    x_position.setElementID(_element.getID());
+
+    for (unsigned ip = 0; ip < n_integration_points; ip++)
+    {
+        x_position.setIntegrationPoint(ip);
+
+        auto const& sm = _shape_matrices[ip];
+        auto &ip_data = _ip_data[ip];
+        auto const& wp = _integration_method.getWeightedPoint(ip);
+        auto const& detJ = ip_data._detJ;
+        auto const& integralMeasure = ip_data._integralMeasure;
+
+        // levelset functions
+        auto const ip_physical_coords = computePhysicalCoordinates(_element, sm.N);
+        double levelsets = calculateLevelSetFunction(fracture_props, ip_physical_coords.getCoords());
+
+        // nodal displacement = u^hat + levelset* [u]
+        NodalDisplacementVectorType nodal_u = nodal_ru + levelsets * nodal_du;
+
+        // strain, stress
+        auto const& B = ip_data._b_matrices;
+        auto const& eps_prev = ip_data._eps_prev;
+        auto const& sigma_prev = ip_data._sigma_prev;
+
+        auto& eps = ip_data._eps;
+        auto& sigma = ip_data._sigma;
+        auto& C = ip_data._C;
+        auto& material_state_variables = *ip_data._material_state_variables;
+
+        eps.noalias() = B * nodal_u;
+
+        if (!ip_data._solid_material.computeConstitutiveRelation(
+                t, x_position, _process_data.dt, eps_prev, eps, sigma_prev,
+                sigma, C, material_state_variables))
+            OGS_FATAL("Computation of local constitutive relation failed.");
+
```

Comment:
It might be good to introduce a constant for detJ * wp.getWeight() * integralMeasure.

Besides,  transpose() seems perform a real transpose. It is true, to place the transpose B matrix with  transposeInPlace() might be good.